synset pos parameter

it176131 opened the pull request:
added `pos` parameter to `.synsets(...)` method so user can filter `wordnet.synsets(...)` directly

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- importing Synset class and added return type hint to __find_synsets(...) method
	- pep8 character limit adjustment

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- added optional pos param to __find_synsets(...) method

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- argument handling for pos param

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- swapping all(map(...)) for set(...).difference(...) which gives a slight boost in speed and readability

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- added try/except to attempt to convert pos arg to list

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- filtering acceptable_pos using pos values and assigning to token_pos which will be used to determine which tokens to get synsets for

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- moved call from self.__synsets declaration into .synsets(...) method allowing user to supply pos args

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- return type hint and docstring for synsets(...) method

it176131 committed:
modified:   tests/test_wordnet_annotator.py
	- added three assertions for pos param in test_english_annotations() method

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- fixed error type hint in synsets(...) method

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- fixed type error in __find_lemmas() method by swapping self.__synsets attribute with self.synsets(...) method
	- pep8 character limit fix in __find_lemmas() method

it176131 committed:
modified:   spacy_wordnet/wordnet_domains.py
	- defined token_synsets as a separate list and filtered returned synsets in wn.synsets and extending token_synsets in __find_synsets(...) method

it176131 committed:
modified:   tests/test_wordnet_annotator.py
	changed expected_adj_synsets to set() instead of {} (a dict) in test_english_annotations() method

it176131 committed:
Merge pull request #1 from it176131/synset_pos_param

Synset pos param

it176131 commented:
Hi @frascuchon 
Any word on when this will be accepted/rejected?

frascuchon commented on spacy_wordnet/wordnet_domains.py:
```diff
@@ -24,17 +26,28 @@ def get_domains_for_synset(synset: Synset) -> List[str]:
 
-    def synsets(self):
-        return self.__synsets
+    def synsets(self, pos: Optional[Union[str,
+                                          List[str]]] = None) -> List[Synset]:
```
```suggestion
    def synsets(self, pos: Optional[Union[str, List[str]]] = None) -> List[Synset]:
```

frascuchon commented on spacy_wordnet/wordnet_domains.py:
```diff
@@ -49,14 +62,43 @@ def wordnet_synsets_for_domain(self, domains: List[str]):
 
     @staticmethod
-    def __find_synsets(token: Token, lang: str):
+    def __find_synsets(token: Token,
+                       lang: str,
+                       pos: Optional[Union[str,
+                                           List[str]]] = None) -> List[Synset]:
```
```suggestion
    def __find_synsets(token: Token,
                       lang: str,
                       pos: Optional[Union[str, List[str]]] = None) -> List[Synset]:
```

frascuchon reviewed (commented):
Great work @it176131 !!!

Just a few code-style suggestions. Once they are resolve, we can merge the PR 

Thanks!

it176131 committed:
Update spacy_wordnet/wordnet_domains.py

param type hint spacing/formatting in synsets(...) method

Co-authored-by: Francisco Aranda <francisco@recogn.ai>

it176131 committed:
Update spacy_wordnet/wordnet_domains.py

param type hint spacing/formatting in __find_synsets(...) method

Co-authored-by: Francisco Aranda <francisco@recogn.ai>

it176131 committed:
Merge branch 'master' into master

it176131 commented:
> Great work @it176131 !!!
> 
> Just a few code-style suggestions. Once they are resolve, we can merge the PR
> 
> Thanks!

Fixed! Thanks for the suggestions 🙂

it176131 committed:
Merge branch 'master' into master

it176131 commented:
> Sorry @it176131
> 
> I've merged some older work. Could you resolve conflicts in your PR?

Resolved!

frascuchon commented on spacy_wordnet/wordnet_domains.py:
```diff
@@ -68,16 +79,43 @@ def wordnet_synsets_for_domain(self, domains: List[str]):
+            token_synsets: List[Synset] = []
+            for p in token_pos:
+                token_synsets.extend(wn.synsets(
+                    word, pos=spacy2wordnet_pos(p), lang=lang
```
@it176131 Is there any reason to not search by the `token.pos` no `pos` param provided?

it176131 replied:
> @it176131 Is there any reason to not search by the token.pos no pos param provided?

Not sure I understand. Are you saying that I can provide a list as an argument to the `spacy2wordnet_pos` func? Or are you asking why I don't limit strictly to the `token.pos`?

If the latter, I wanted the user to have the ability to filter to multiple `pos`, not just the `token.pos`. As an example, if the `token.pos` is `VERB`, but the user wants `VERB` and `NOUN`, then I'd need to provide both options rather than the single.

Of course if the `spacy2wordnet_pos` function can take a list, then I can do away with the loop. But I didn't think it could.

Does that answer the question or am I missing the mark?

frascuchon replied:
What I mean is that, if the user does not provide the `pos` parameter, the wordnet search will use all available pos for finding synset.

My proposal is just to keep the old behavior and set as default the `token.pos` value.

I will add a code suggestion to implement my proposal.

frascuchon commented on spacy_wordnet/wordnet_domains.py:
```diff
@@ -68,16 +79,43 @@ def wordnet_synsets_for_domain(self, domains: List[str]):
 
     @staticmethod
-    def __find_synsets(token: Token, lang: str):
+    def __find_synsets(token: Token,
+                       lang: str,
+                       pos: Optional[Union[str, List[str]]] = None) -> List[Synset]:
+        if pos is None:
+            pos = ["verb", "noun", "adj"]
+
+        elif isinstance(pos, str):
+            pos = [pos]
+
+        elif not isinstance(pos, list):
+            try:
+                pos = list(pos)
+
+            except TypeError:
+                raise TypeError("pos argument must be None, type str, or type"
+                                "list.")
+
+        acceptable_pos = {"verb": VERB, "noun": NOUN, "adj": ADJ}
+
+        # check if any element in `pos` is not in `acceptable_pos`
+        if set(pos).difference(acceptable_pos):
+            raise ValueError("pos argument must be a combination of 'verb', "
+                             "'noun', or 'adj'.")
+
+        token_pos: List[int] = [acceptable_pos[k] for k in pos]
```
```suggestion
    def __find_synsets(token: Token,
                       lang: str,
                       pos: Optional[Union[str, List[str]]] = None) -> List[Synset]:
        if pos is None:
            pos = []
        elif isinstance(pos, str):
            pos = [pos]
        elif not isinstance(pos, list):
            try:
                pos = list(pos)
            except TypeError:
                raise TypeError("pos argument must be None, type str, or type list.")
        
        acceptable_pos = {"verb": VERB, "noun": NOUN, "adj": ADJ} # We can define this as a private class constant
        # check if any element in `pos` is not in `acceptable_pos`
        if set(pos).difference(acceptable_pos):
            raise ValueError("pos argument must be a combination of 'verb', "
                             "'noun', or 'adj'.")

        token_pos: List[int] = [acceptable_pos[k] for k in pos]
        if not token_pos:
           token_pos = [token.pos]
```

it176131 replied:
Okay I understand now! I can get behind this suggestion. I'm going to update the docstring and unit test.

it176131 committed:
use token.pos if pos argument is none to mimic previous behavior.

Co-authored-by: Francisco Aranda <francisco@recogn.ai>

it176131 committed:
Update wordnet_domains.py

modified docstring to reflect what happens if pos argument is none

it176131 committed:
modified:   tests/test_wordnet_annotator.py
	- added assert to test that list of pos args will return expected results

it176131 committed:
Merge branch 'master' of https://github.com/it176131/spacy-wordnet

it176131 committed:
modified:   tests/test_wordnet_annotator.py
	- added test for when pos argument is none

it176131 commented:
@frascuchon I've updated the docstring and added a couple assertion checks in the the unit test (one for when `pos` arg is a list of values and one for when the arg is `None`).

Waiting for your review

frascuchon commented on spacy_wordnet/wordnet_domains.py:
```diff
@@ -68,16 +79,40 @@ def wordnet_synsets_for_domain(self, domains: List[str]):
+           token_pos = [token.pos]
         word_variants = [token.text]
-        if token.pos in [VERB, NOUN, ADJ]:
+        if token.pos in token_pos:
```
I've seen a problem here:

If no `pos` param is provided, and `token.pos` is used as default, the `token.lemma_` is always included as a word variant. This could cause a weird behavior of the component

```suggestion
        if token.pos in (token_pos if pos else acceptable_pos.values()):
```

it176131 replied:
Let me make sure I understand correctly.

If `pos is None` and `token.pos` is something other than `NOUN`, `VERB`, or `ADJ` --> this may cause unexpected `token.lemma_` values to be appended to the `word_variants` list.

# EDIT
I tested with `token = nlp("quickly")[0]` before implementing your suggestion:

``` python
token = nlp("quickly")[0]

synsets = token._.wordnet.synsets(pos=None)

set(synsets)

>>> {Synset('cursorily.r.01'), Synset('promptly.r.01'), Synset('quickly.r.01')}
```

I agree this is unexpected.

it176131 committed:
Update spacy_wordnet/wordnet_domains.py

Checking if `token.pos` is an acceptable value before appending its lemma to the `word_variants` list. This avoids unexpected results such as when `token.pos` is an `ADVERB`.

Co-authored-by: Francisco Aranda <francisco@recogn.ai>

it176131 commented:
@frascuchon I'm going to update the docstring to reflect the recent changes and add an edge case to the unit test.

it176131 committed:
Update wordnet_domains.py

Updated docstring so user knows results are limited to NOUN, VERB, and ADJ even if `pos` is None.
